copy.go

```diff
@@ -54,6 +56,41 @@ type CopyOptions struct {
 	MapRoot func(ctx context.Context, src content.Storage, root ocispec.Descriptor) (ocispec.Descriptor, error)
 }
 
+// selectPlatform implements platform filter and returns the descriptor of
+// the first matched manifest from the manifest list / image index
+func (o *CopyOptions) selectPlatform(ctx context.Context, src content.Storage, root ocispec.Descriptor, p *ocispec.Platform) (ocispec.Descriptor, error) {
+	if root.MediaType == docker.MediaTypeManifestList || root.MediaType == ocispec.MediaTypeImageIndex {
+		manifests, err := content.Successors(ctx, src, root)
+		if err != nil {
+			return ocispec.Descriptor{}, errdef.ErrNotFound
+		}
+
+		// platform filter
+		for _, m := range manifests {
+			matched := platform.MatchPlatform(m.Platform, p)
+			if matched {
+				return m, nil
+			}
+		}
+		return ocispec.Descriptor{}, errdef.ErrNotFound
+	}
+	return ocispec.Descriptor{}, errdef.ErrNotFound
+}
+
+// AddPlatformFilter adds the selectPlatform func into the MapRoot func
+func (o *CopyOptions) AddPlatformFilter(p *ocispec.Platform) {
+	mapRoot := o.MapRoot
+	o.MapRoot = func(ctx context.Context, src content.Storage, root ocispec.Descriptor) (ocispec.Descriptor, error) {
+		var err error
+		if mapRoot != nil {
+			if root, err = mapRoot(ctx, src, root); err != nil {
```

Comment:
one question: is it possible that `root` gets changed once returned?

Reply:
According to line 141 in copy.go https://github.com/oras-project/oras-go/blob/bc29e42ccc18897b95d1162f6aeb9ff4e3635da0/copy.go#L141
The root will eventually be changed to the returned descriptor if there is the matched manifest

Reply:
@binbin-li The root will also being used for tagging before return